File: kora-app-annotation-processor/src/main/java/ru/tinkoff/kora/kora/app/annotation/processor/ServiceTypesHelper.java
package ru.tinkoff.kora.kora.app.annotation.processor;

import ru.tinkoff.kora.annotation.processor.common.CommonClassNames;

import javax.lang.model.element.TypeElement;
import javax.lang.model.type.DeclaredType;
import javax.lang.model.type.TypeMirror;
import javax.lang.model.util.Elements;
import javax.lang.model.util.Types;
import java.util.Objects;

public class ServiceTypesHelper {
    private final Elements elements;
    private final Types types;
    private final TypeElement wrappedTypeElement;
    private final DeclaredType wrappedType;
    private final TypeElement interceptorTypeElement;
    private final DeclaredType interceptorType;
    private final TypeMirror lifecycleType;

    public ServiceTypesHelper(Elements elements, Types types) {
        this.elements = elements;
        this.types = types;
        this.wrappedTypeElement = Objects.requireNonNull(this.elements.getTypeElement(CommonClassNames.wrapped.canonicalName()));
        this.wrappedType = Objects.requireNonNull(this.types.getDeclaredType(this.wrappedTypeElement, this.types.getWildcardType(null, null)));
        this.interceptorTypeElement = Objects.requireNonNull(this.elements.getTypeElement(CommonClassNames.graphInterceptor.canonicalName()));
        this.interceptorType = Objects.requireNonNull(this.types.getDeclaredType(this.interceptorTypeElement, this.types.getWildcardType(null, null)));
        var lifecycleTypeElement = Objects.requireNonNull(this.elements.getTypeElement(CommonClassNames.lifecycle.canonicalName()));
        this.lifecycleType = Objects.requireNonNull(lifecycleTypeElement.asType());
    }

    public boolean isAssignableToUnwrapped(TypeMirror maybeWrapped, TypeMirror typeMirror) {
        if (!this.types.isAssignable(maybeWrapped, this.wrappedType)) {
            return false;
        }
        var wrappedParameterElement = this.wrappedTypeElement.getTypeParameters().get(0); // somehow it can be changed during execution
        var declaredType = (DeclaredType) maybeWrapped;
        var unwrappedType = this.types.asMemberOf(declaredType, wrappedParameterElement);
        return this.types.isAssignable(unwrappedType, typeMirror);
    }

    public boolean isSameToUnwrapped(TypeMirror maybeWrapped, TypeMirror typeMirror) {
        if (!this.types.isAssignable(maybeWrapped, this.wrappedType)) {
            return false;
        }
        var wrappedParameterElement = this.wrappedTypeElement.getTypeParameters().get(0); // somehow it can be changed during execution
        var declaredType = (DeclaredType) maybeWrapped;
        var unwrappedType = this.types.asMemberOf(declaredType, wrappedParameterElement);
        return this.types.isSameType(unwrappedType, typeMirror);
    }

    public boolean isInterceptorFor(TypeMirror maybeInterceptor, TypeMirror typeMirror) {
        if (!this.types.isAssignable(maybeInterceptor, this.interceptorType)) {
            return false;
        }
        var interceptorTypeParameter = this.interceptorTypeElement.getTypeParameters().get(0); // somehow it can be changed during execution
        var declaredType = (DeclaredType) maybeInterceptor;
        var interceptedType = this.types.asMemberOf(declaredType, interceptorTypeParameter);
        return this.types.isSameType(interceptedType, typeMirror);
    }

    public boolean isInterceptor(TypeMirror maybeInterceptor) {
        return this.types.isAssignable(maybeInterceptor, this.interceptorType);
    }

    public TypeMirror getInterceptedType(TypeMirror maybeInterceptor) {
        var interceptorTypeParameter = this.interceptorTypeElement.getTypeParameters().get(0); // somehow it can be changed during execution
        var declaredType = (DeclaredType) maybeInterceptor;
        return this.types.asMemberOf(declaredType, interceptorTypeParameter);
    }

    public boolean isLifecycle(TypeMirror type) {
        return this.types.isAssignable(type, this.lifecycleType) || this.isAssignableToUnwrapped(type, this.lifecycleType);
    }
}

File: kora-app-symbol-processor/src/main/kotlin/ru/tinkoff/kora/kora/app/ksp/ServiceTypesHelper.kt
package ru.tinkoff.kora.kora.app.ksp

import com.google.devtools.ksp.getAllSuperTypes
import com.google.devtools.ksp.getDeclaredFunctions
import com.google.devtools.ksp.processing.Resolver
import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.google.devtools.ksp.symbol.KSType
import ru.tinkoff.kora.ksp.common.CommonClassNames


class ServiceTypesHelper(val resolver: Resolver) {
    private val interceptorClassDeclaration = resolver.getClassDeclarationByName(resolver.getKSNameFromString(CommonClassNames.graphInterceptor.canonicalName))!!
    private val interceptorType = interceptorClassDeclaration.asStarProjectedType()
    private val interceptorInitFunction = interceptorClassDeclaration.getDeclaredFunctions()
        .filter { it.simpleName.asString() == "init" && it.parameters.size == 1 }
        .first()

    private val lifecycleClassDeclaration = resolver.getClassDeclarationByName(resolver.getKSNameFromString(CommonClassNames.lifecycle.canonicalName))!!
    private val lifecycleType = lifecycleClassDeclaration.asStarProjectedType()

    private val wrappedClassDeclaration = resolver.getClassDeclarationByName(resolver.getKSNameFromString(CommonClassNames.wrapped.canonicalName))!!
    private val wrappedType = wrappedClassDeclaration.asStarProjectedType()
    private val wrappedValueFunction = wrappedClassDeclaration.getDeclaredFunctions()
        .filter { it.simpleName.asString() == "value" && it.parameters.isEmpty() }
        .first()


    fun isAssignableToUnwrapped(maybeWrapped: KSType, type: KSType): Boolean {
        if (!wrappedType.isAssignableFrom(maybeWrapped)) {
            return false
        }
        val maybeWrappedDeclaration = maybeWrapped.declaration as KSClassDeclaration
        val wrappedClassDeclaration = maybeWrappedDeclaration.getAllSuperTypes().plus(sequence { this.yield(maybeWrappedDeclaration.asType(listOf())) })
            .first { CommonClassNames.wrapped.canonicalName == it.declaration.qualifiedName?.asString() }
            .declaration as KSClassDeclaration
        val wrappedValueFunction = wrappedClassDeclaration.getAllFunctions()
            .filter { it.simpleName.asString() == "value" }
            .first()
        val unwrappedType = wrappedValueFunction.asMemberOf(maybeWrapped).returnType!!
        return type.isAssignableFrom(unwrappedType)
    }

    fun isSameToUnwrapped(maybeWrapped: KSType, type: KSType): Boolean {
        if (!wrappedType.isAssignableFrom(maybeWrapped)) {
            return false
        }
        val unwrappedType = wrappedValueFunction.asMemberOf(maybeWrapped).returnType!!
        return unwrappedType.makeNotNullable() == type // platform nullability ruins equality
    }

    fun isInterceptorFor(maybeInterceptor: KSType, type: KSType): Boolean {
        if (!interceptorType.isAssignableFrom(maybeInterceptor)) {
            return false
        }

        val interceptsType = interceptorInitFunction.asMemberOf(maybeInterceptor).parameterTypes[0]!!
        return type == interceptsType.makeNotNullable()
    }


    fun interceptType(maybeInterceptor: KSType): KSType {
        if (!interceptorType.isAssignableFrom(maybeInterceptor)) {
            throw IllegalArgumentException()
        }

        return interceptorInitFunction.asMemberOf(maybeInterceptor).parameterTypes[0]!!.makeNotNullable()
    }

    fun isLifecycle(type: KSType): Boolean {
        return lifecycleType.isAssignableFrom(type) || isAssignableToUnwrapped(type, this.lifecycleType)
    }

    fun isInterceptor(type: KSType) = interceptorType.isAssignableFrom(type)
}
